grub2_pvgrub_image, grub: Refactor, add PVH support

hehongbo opened the pull request:
This one is revived from #121514 by @radhus, with the GRUB configuration files and memdisk stuff being reused from the `grub2_pvgrub_image` instead of duplicated, since the making of PvGrub and PvhGrub image shares the same logic.

The result, is a new `grub2_pvhgrub_image` being provided along with the old `grub2_pvgrub_image`, which can be used to boot PVH guests. And this, hopefully, provides an upgrade path for PV guests since PV is old, and has performance overhead for historical reasons.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

hehongbo commented:
ping @NixOS/xen-project 

I still have to iron out these errors from format checks, meanwhile, as this is also discussed on William's original PR, should we skip the arm/aarch64 architectures for now? I can't imagine if PvhGrub is useful here in Nixpkgs if we don't have the hypervisor itself working properly. (Maybe after #345217?)

SigmaSquadron commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,65 @@
+
+    buildInputs = lib.optional buildPvImage grub2_xen ++ lib.optional buildPvhImage grub2_xen_pvh;
+
+    buildCommand =
```
This really needs to be split into a `{configure,build,install}Phase`. Overriding the Standard Environment `buildCommand` is a really bad idea.

SigmaSquadron commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,65 @@
+        }-${grubVariant}.bin" $out/lib/grub-${grubVariant}/
+      '';
+
+    meta = with lib; {
```
```suggestion
    meta = {
```

SigmaSquadron commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,65 @@
+  grub2_xen_pvh,
+}:
+
+assert (buildPvImage != buildPvhImage);
```
It might be best to have a single `imageType` attribute and an assertion that allows you to pick between `pv` and `pvh`. See [`lib.asserts.assertOneOf`](https://noogle.dev/f/lib/asserts/assertOneOf) for more information.
This will allow you to pick a default image type, and remove one of the explicit attributes in `all-packages.nix`.

hehongbo replied:
That's indeed better, I was fighting myself about the idea that only one boolean flag is enough to tell PV versus PVH while I'm using 2 (`buildPvImage` and `buildPvhImage`) for expressiveness. 

That also remove sthe current building failure since I was blocking the direct evaluation of the intermediate `grub2-xen-images` package via assertion.

However, this will make the intermediate `grub2-xen-images` package available at search.nixos.org, will this lead to misunderstandings or are we fine with it? And if we are, which one should be the default? I'm going with PVH since it's a recommended option for existing PV users and usecases, let me know if I should go with the other way around.

hehongbo replied:
By the way, should `grub2-xen-images` be renamed into `grub2_xen_images`? All other GRUB2 packages use underscores, so I'm feeling awkward after my initial commit.

SigmaSquadron replied:
> However, this will make the intermediate grub2-xen-images package available at search.nixos.org, will this lead to misunderstandings or are we fine with it?

It's fine. We also display `python3Packages.xen`, which is an intermediate package.

SigmaSquadron replied:
> which one should be the default?

As you said, PVH is the recommended option.

SigmaSquadron replied:
> By the way, should `grub2-xen-images` be renamed into `grub2_xen_images`? All other GRUB2 packages use underscores, so I'm feeling awkward after my initial commit.

I'll leave it to your discretion. I don't mind one way or the other.

SigmaSquadron commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,65 @@
+        # We include all modules except all_video.mod as otherwise grub will fail printing "no symbol table"
+        # if we include it.
+        grub-mkimage \
+          -O "${efiSystemsBuild.${stdenv.hostPlatform.system}.target}-${grubVariant}" \
```
```suggestion
          -O "${efiSystemsBuild.${stdenv.buildPlatform.system}.target}-${grubVariant}" \
```
As far as I understand it, we can cross-compile by choosing a different system, right? We aren't actually executing any built images.

hehongbo replied:
This is interesting and makes sense, I don't know how but I believe it's possible to generate the image from a different architecture where `grub-mkimage` runs, as long as the corresponding assets (like mods) are ready.

SigmaSquadron commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,65 @@
+          -O "${efiSystemsBuild.${stdenv.hostPlatform.system}.target}-${grubVariant}" \
+          -c grub-bootstrap.cfg \
+          -m memdisk.tar \
+          -o "grub-${efiSystemsBuild.${stdenv.hostPlatform.system}.target}-${grubVariant}.bin" \
```
```suggestion
          -o "grub-${imageName}.bin" \
```
Since `${efiSystemsBuild.${stdenv.hostPlatform.system}.target}-${grubVariant}` is repeated quite a few times, add it to the `let...in` expression above.

hehongbo replied:
Oh yes, 4 times. I should have noticed this when reusing the old `grub2_pvgrub_image` package.


I think we should call it `grubFormat`, as it's seen in `grub-mkimage(1)` man page, this reflects how the upstream calls the "target_cpu-platform" argument.

SigmaSquadron commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,65 @@
+          -c grub-bootstrap.cfg \
+          -m memdisk.tar \
+          -o "grub-${efiSystemsBuild.${stdenv.hostPlatform.system}.target}-${grubVariant}.bin" \
+          $(ls "${if buildPvImage then grub2_xen else grub2_xen_pvh}/lib/grub/${
```
I'm not sure how I feel about relying on the output of `ls`. Can this be changed for a native bash glob?

hehongbo replied:
Guess what, while I'm testing if I correctly excluded the `all_video.mod` via bash glob I found it does not exist anymore. It exists in `grub2` but not `grub2_xen`.

The original logic that excluding `all_video.mod` was implemented 8 years ago, and I believe the upstream has already ironed that out.

See [6371e9c10433578bb236a8284ddb9ce9e201eb59](https://git.savannah.gnu.org/cgit/grub.git/commit/?id=6371e9c10433578bb236a8284ddb9ce9e201eb59) and [a40b219e269ac407c10a569ce6b0c5d7f419d127](https://git.savannah.gnu.org/cgit/grub.git/commit/?id=a40b219e269ac407c10a569ce6b0c5d7f419d127)

SigmaSquadron commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,65 @@
+        Paravirtualized (PV) or Paravirtualized Hardware (PVH) Xen guests
+      '';
+
+      platforms = platforms.gnu ++ platforms.linux;
```
```suggestion
      platforms = lib.platforms.linux;
```

SigmaSquadron commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,65 @@
+        This package provides PvGrub/PvhGrub images for booting
+        Paravirtualized (PV) or Paravirtualized Hardware (PVH) Xen guests
+      '';
+
```
```suggestion
  maintainers = lib.teams.xen.members;
```

hehongbo commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,65 @@
+  stdenv.mkDerivation rec {
+    name = "grub2-xen-images";
+
+    configs = ./configs;
```
I'm not sure if GRUB configurations should be considered "sources" since they're not source code in the traditional sense. As for `env`, it gives me an error because it only allows derivation, string, boolean or integer attributes, so let me know if going with this is proper.

```nix
env = {
  GRUB_CONFIGS = "${./configs}";
};
```

Or, if we decide to give the `configs` as string as env, should we also put that `${efiSystemsBuild.${stdenv.hostPlatform.system}.target}-${grubVariant}` in env as `GRUB_FORMAT` or such?

SigmaSquadron replied:
If `GRUB_CONFIGS = "${./configs}";` works, then I would recommend that.

hehongbo commented on pkgs/by-name/gr/grub2-xen-images/package.nix:
```diff
@@ -0,0 +1,68 @@
+      platforms = lib.platforms.linux;
+    };
+  })
+)
```
Both parenthesis are gone, and since `configs` went into `env` as `GRUB_CONFIGS`, we won't even need `(finalAttrs: {` then.

SigmaSquadron commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3550,6 +3550,16 @@ with pkgs;
+    xenPvhSupport = true;
+  };
+
+  grub2_pvgrub_image = callPackage ../by-name/gr/grub2_xen_images/package.nix {
+    grubPlatform = "xen";
+  };
+
+  grub2_pvhgrub_image = callPackage ../by-name/gr/grub2_xen_images/package.nix {};
```
Instead of this, I'd recommend moving `grub2_xen_images/package.nix` to `grub2_pvhgrub_image/package.nix`. It seems really weird to have three packages for this, and one being just a renamed version of another.

hehongbo replied:
done, and does the package description feel out of place to you then? Or should I provide 2 variants of it based on `grubPlatform`?

SigmaSquadron replied:
Yeah, might be a good idea to have the `description` and `longDescription` change depending on the target.

hehongbo replied:
done.

Despite similarities, I ended up not liking the large pile of `lib.optionalString` stacked together, so I made them 2 separate strings.🫠

SigmaSquadron commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3550,6 +3550,16 @@ with pkgs;
+    xenPvhSupport = true;
+  };
+
+  grub2_pvgrub_image = callPackage ../by-name/gr/grub2_xen_images/package.nix {
+    grubPlatform = "xen";
+  };
```
This should probably be an override instead.

hehongbo replied:
done

SigmaSquadron commented on pkgs/top-level/all-packages.nix:
```diff
@@ -3550,6 +3550,14 @@ with pkgs;
+    xenPvhSupport = true;
+  };
+
+  grub2_pvgrub_image = (callPackage ../by-name/gr/grub2_pvhgrub_image/package.nix { }).override {
```
```suggestion
  grub2_pvgrub_image = grub2_pvhgrub_image.override {
```

SigmaSquadron reviewed (approved):
Looks good to me! Thanks for picking this up.

hehongbo committed:
grub: build grub2_pvhgrub_image together with grub2_pvgrub_image

This enables the GRUB package to be built for the `xen_pvh` platform, which is
needed to generate a corresponding GRUB image and boot PVH DomUs.

Also, modifications made to the original `grub2_pvgrub_image` package:
- Use split phases and avoid overriding `buildCommand` during the build.
- Remove logic that excludes `all_video.mod` (already ironed out by upstream).
- Fix `lib.platforms` and remove platforms that cannot build.
- Update the description.
- Add maintainer information of the Xen Project team.

Signed-off-by: Hongbo <hehongbo@mail.com>
Reviewed-by: Fernando Rodrigues <alpha@sigmasquadron.net>
Co-authored-by: William Johansson <radar@radhuset.org>

hehongbo commented:
No changes; just rebase and resolve conflicts. Should wait for #395558 and do another rebase, probably.

hehongbo commented:
No changes; just rebase and resolve conflicts.

hehongbo commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 374753`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 4 packages built:</summary>
  <ul>
    <li>grub2_pvgrub_image</li>
    <li>grub2_pvhgrub_image</li>
    <li>grub2_xen_pvh</li>
    <li>grub2_xen_pvh.debug</li>
  </ul>
</details>

SigmaSquadron commented on pkgs/by-name/gr/grub2_pvhgrub_image/package.nix:
```diff
@@ -0,0 +1,87 @@
+          explanations.
+        '';
+
+    maintainers = lib.teams.xen.members;
```
```suggestion
    teams = [ lib.teams.xen ];
```
This has been so long without being merged that there's a new `meta` attribute we should be using. 🙃

hehongbo committed:
grub: build grub2_xen_pvh

This enables the GRUB package to be built for the `xen_pvh` platform, which is
needed to generate a corresponding GRUB image and boot PVH DomUs.

Signed-off-by: Hongbo <hehongbo@mail.com>
Co-authored-by: William Johansson <radar@radhuset.org>

hehongbo commented:
Rebase again to resolve conflicts, and I realized that I should fix platform support declarations/meta after considering the actual support matrix, which was #427180 and #428878 that had prompted me to do so.

- `xenPvhSystemsBuild` will be used to present `meta.platforms` when building `pkgs.grub2_xen_pvh`. It was previously set to `platforms.gnu` plus `platforms.linux`, which is not accurate. The #428878 fixed that for `pkgs.grub2_xen` (PvGrub package), and now we just align with that.
- And once again, I renamed `xenSystemsBuild` into `targets` in `grub2_pvhgrub_image`, similar to William's original PR. Because now we have `xenSystemsBuild` and `xenPvhSystemsBuild` separately in the GRUB main package, we should either align with that, or move the attrset's name to something else to avoid future confusion.
- And similarly, chain `meta.platforms` to the support `targets` attrset in `grub2_pvhgrub_image`. According to [configure.ac](https://cgit.git.savannah.gnu.org/cgit/grub.git/tree/configure.ac), only the two x86 platforms are supported in PvGrub and PvhGrub, with the latter one sharing the `i386` binary/image for both platforms.

It's mysterious that `arm`, `aarch64`, `riscv32`, `riscv64` were in `efiSystemsBuild` of `pkgs.grub2_pvgrub_image` (and by the way, the "efi" is because of copy-pasted code and has nothing to do with EFI). It turned out to be a fix of evaluation errors in #213612, all because that `meta.platforms` was set incorrectly in the first place. Now we should make it clear, only two x86 platforms are supported, ARM and RISC-V are not, at least for now.
